bpfctl/src/main.rs

```diff
@@ -43,7 +43,21 @@ enum Commands {
     /// Unload a BPF program using the UUID.
     Unload(UnloadArgs),
     /// List all BPF programs loaded via bpfd.
-    List,
+    List(ListArgs),
+    /// Get a program's metadata information specified by kernel id.
+    Get { kernel_id: u32 },
+}
+
+#[derive(Args)]
+struct ListArgs {
+    /// Optional: Program type to list
+    /// Example: --program-type xdp
+    #[clap(short, long, verbatim_doc_comment)]
+    program_type: Option<ProgramType>,
+
```

Comment:
```suggestion
    /// Optional: Program type to list
    /// Example: --program-type xdp
    /// [possible values: unspec, socket-filter, kprobe, tc, sched-act,
    ///                   tracepoint, xdp, perf-event, cgroup-skb,
    ///                   cgroup-sock, lwt-in, lwt-out, lwt-xmit, sock-ops,
    ///                   sk-skb, cgroup-device, sk-msg, raw-tracepoint,
    ///                   cgroup-sock-addr, lwt-seg6-local, lirc-mode2,
    ///                   sk-reuseport, flow-dissector, cgroup-sysctl,
    ///                   raw-tracepoint-writable, cgroup-sockopt, tracing,
    ///                   struct-ops, ext, lsm, sk-lookup, syscall]
    #[clap(short, long, verbatim_doc_comment, hide_possible_values=true)]
    program_type: Option<ProgramType>,

```

Not a fan of the original helptext output (wraps on my terminal):
```
$ bpfctl list --help
List all BPF programs loaded via bpfd

Usage: bpfctl list [OPTIONS]

Options:
  -p, --program-type <PROGRAM_TYPE>  Optional: Program type to list
                                     Example: --program-type xdp [possible values: unspec, socket-filter, kprobe, tc, sched-a
ct, tracepoint, xdp, perf-event, cgroup-skb, cgroup-sock, lwt-in, lwt-out, lwt-xmit, sock-ops, sk-skb, cgroup-device, sk-msg,
 raw-tracepoint, cgroup-sock-addr, lwt-seg6-local, lirc-mode2, sk-reuseport, flow-dissector, cgroup-sysctl, raw-tracepoint-wr
itable, cgroup-sockopt, tracing, struct-ops, ext, lsm, sk-lookup, syscall]
  -a, --all                          
  -h, --help                         Print help
```
With this change, it now looks like:
```
$ bpfctl list --help
List all BPF programs loaded via bpfd

Usage: bpfctl list [OPTIONS]

Options:
  -p, --program-type <PROGRAM_TYPE>  Optional: Program type to list
                                     Example: --program-type xdp
                                     [possible values: unspec, socket-filter, kprobe, tc, sched-act,
                                                       tracepoint, xdp, perf-event, cgroup-skb,
                                                       cgroup-sock, lwt-in, lwt-out, lwt-xmit, sock-ops,
                                                       sk-skb, cgroup-device, sk-msg, raw-tracepoint,
                                                       cgroup-sock-addr, lwt-seg6-local, lirc-mode2,
                                                       sk-reuseport, flow-dissector, cgroup-sysctl,
                                                       raw-tracepoint-writable, cgroup-sockopt, tracing,
                                                       struct-ops, ext, lsm, sk-lookup, syscall]
  -a, --all                          
  -h, --help                         Print help
```

I know there is more maintenance with this approach, but it makes it so much easier to read. And I don't think the kernel ProgramType will change that often, and if it does and the helptext part is missed, it will still filter properly on the new value, just the list will be outdated.

Reply:
Sounds good to me, I don't think there's much more maintenance I doubt many bpf program types will be added in the future.